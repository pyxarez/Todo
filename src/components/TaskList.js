import React from 'react';

import TextBox from './TextBox';
import './TaskList.res/style.css';

TaskList.propTypes = {
  children: React.PropTypes.oneOfType(
    [
      React.PropTypes.arrayOf(React.PropTypes.element.isRequired),
      React.PropTypes.bool,
      React.PropTypes.object
    ]
  ),
  onClick: React.PropTypes.func.isRequired
}

export default function TaskList({ onClick, children }) {
  return (
    <div className="my-tasklist-component">
      <TextBox
        placeholder="Enter task title"
        onClick={onClick}/>
      <div className="tasklist">
        {children}
      </div>
    </div>
  );
}


