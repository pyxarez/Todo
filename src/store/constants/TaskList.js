export const ADD_TASK = 'ADD_TASK';
export const CHECK_DONE = 'CHECK_DONE';
export const GET_TASKS = 'GET_TASKS';
export const SAVE_TASK_CHANGES = 'SAVE_TASK_CHANGES';
