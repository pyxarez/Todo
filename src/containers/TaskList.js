import React, { Component, PropTypes } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { addTask, getTasks } from '../store/actions/TaskListActions';
import { getProgress } from '../store/actions/ProgressActions';

import TaskList from '../components/TaskList';
import TaskContainer from './Task';

export class TaskListContainer extends Component {
  static propTypes = {
    showDone: PropTypes.bool.isRequired,
    tasks: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.node.isRequired,
      title: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
      description: React.PropTypes.string.isRequired
    })),
    addTask: PropTypes.func.isRequired,
    getTasks: PropTypes.func.isRequired,
    getProgress: React.PropTypes.func.isRequired,
    URLParams: PropTypes.shape({
      id: PropTypes.number,
      category: PropTypes.string,
      filter: PropTypes.string
    })
  }

  handleAddNewTaskClick = (title) => {
    const {
      addTask,
      getProgress
    } = this.props;
    const id = this.props.URLParams.id;

    return addTask(id, title)
      .then(() => { getProgress();})
      .catch(e => { alert(e); });
  }

  componentDidMount = () => {
    const { id } = this.props.URLParams;
    
    if (typeof id !== 'undefined') {
      this.props.getTasks(id);
    }
  }

  componentWillReceiveProps = (nextProps) => {
    const nextId = nextProps.URLParams.id;
    const prevId = this.props.URLParams.id

    if ( (typeof nextId !== 'undefined') && (nextId !== prevId) ){
      this.props.getTasks(nextId);
    }
  }

  _createTasks = () => {
    const {
      showDone,
      tasks
    } = this.props;
    const {
      id: categoryId,
      filter: keyWord = ''
    } = this.props.URLParams;

    const sortedTaskList = showDone
      ? tasks
      : tasks.filter(task => !task.isDone);

    return sortedTaskList
      .filter((task) => task.title
        .toLowerCase()
        .includes(keyWord
          .toLowerCase()
          .trim()
         )
       )
      .map((task, index) => (
          <TaskContainer
            key={task.id}
            categoryId={categoryId}
            task={task}/>
        )
      );
  }

  render() {
    const { tasks } = this.props;
    return (
      <TaskList onClick={this.handleAddNewTaskClick}>
        {tasks.length > 0 && this._createTasks()}
      </TaskList>
    );
  }
}

const mapStateToProps = ({ tasks }) => {
  return {
    tasks
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addTask: bindActionCreators(addTask, dispatch),
    getTasks: bindActionCreators(getTasks, dispatch),
    getProgress: bindActionCreators(getProgress, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TaskListContainer);