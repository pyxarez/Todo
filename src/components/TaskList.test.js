import React from 'react';
import { shallow } from 'enzyme';

import TaskList from './TaskList';

const setup = () => {
  const props = {
    onClick: () => {}
  }
  const wrapper = shallow(
    <TaskList {...props}>
      <div className="testClass" />
    </TaskList>
  );

  return {
    props,
    wrapper
  }
}

describe('Components :: TaskList', () => {
  it('should render self', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});