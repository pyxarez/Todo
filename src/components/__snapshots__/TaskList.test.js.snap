exports[`Components :: TaskList should render self 1`] = `
<div
  className="my-tasklist-component">
  <TextBox
    isFocused={false}
    onClick={[Function]}
    placeholder="Enter task title" />
  <div
    className="tasklist">
    <div
      className="testClass" />
  </div>
</div>
`;
